doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,231 @@
+
+Source frontier differences are used in some form in the metrics proposed in [Snapshot Progress](#snapshot-progress) and
+[Steady-state](#steady-state).
+
+#### Kafka frontier differences
+
+For example, consider the following 2 kafka source frontiers:
+```
+# Partition Id: Offset
+f1: {0: 100, 1: 150}
+f2: {0: 150, 1: 170}
+```
+
+In this case, the source frontier difference is `f2 - f1 = 70 offsets`. Similarly:
+
+```
+{0: 100, 1: 100} - {0: 0} = 200 offsets
+```
+
+#### Postgres frontier differences
+
+Postgres has two separate definitions:
+
+For the number of rows in each table:
+```
+{table1: 110 rows, table2: 15 rows} - {table1: 100 rows, table2: 10 rows} = 15 rows
+```
+
+And for lsn's during replication:
+```
+100100 lsn - 100050 lsn = 50 bytes
+```
+
+#### MySQL frontier differences
+
+Similar to Postgres, mysql has a rows-per-table difference, as well as a replication difference.
+
+However, during replication, MySQL's source timestamp is partially ordered, instead of just a single
+lsn number:
+
+```
+# source id: gtid
+f1: {0: 1000th gtid, 1: 0th gtid}
+f2: {0: 1010th gtid, 1: 10th gtid}
+```
+In this case, `f2 - f1 = 20 transactions`
+
+#### Comparisons
+
+As you can see, source frontier differences depend on the source, and are primarily useful for _calculating rates_ and when
+_comparing values_. For example:
+
+- If a source's frontier has moved from `{0: 10}` to `{0: 70}` in 1 minute, Materialize can be said to have processed 1 offset/minute.
+- If a source's frontier snapshot frontier is `{table1: 100 rows}` and its current frontier is `{table1: 50}`, Materialize can be said
+to have processed about 50% of the snapshot.
+
+Additionally, differences are designed to be _lower-bounded estimates_. For example, `{0: 100} - {0: 0, 1: 50}` is 100, as we
+choose the lower bound where we assume no offsets past 50 for partition 1 exist.
+
+
+## Snapshot Progress
+
+The first set of metrics this design document proposes involve _snapshot progress_, which is a _lower bound_ on the _percentage_ of the source's snapshot Materialize
+has _read_. These metrics are designed to answer #1 in [the problem statement](#the-problem).
+
+We will introduce 2 new columns in `mz_source_statistics_per_worker`:
+
+```
+| `snapshot_total`     | [`uint8`] | The total number of upstream values that are part of the snapshot. |
+| `snapshot_progress   | [`uint8`] | The number of upstream values Materialize has read so far.         |
+```
+
+The unit of _values_ depends on the source type, and will be _rows_ for MySQL and Postgres, and _offsets_ for kafka.
+
+These values can be summed across workers and compared (`snapshot_progress / snapshot_progress - snapshot_total`) to produce
+a _lower-bound_ estimate on the progress we have made reading the source's snapshot.
+
+### Source frontier difference
+
+Note that this percentage is calculated using _source frontier difference_, between the `snapshot_total` and `snapshot_progress`.
```

Comment:
This is a bit misleading since for MySQL & Postgres the frontiers are in terms of LSNs / transactions but `snapshot_total` and `snapshot_progress` will be in terms of row-counts

Reply:
I tried to capture that there are 2 types of frontiers for pg and mysql in the above sections, is there more context I should add?

Reply:
Actually, I see your point, ill adjust the structure a bit